File: web/src/core/extraction/InvoiceFieldParser.ts
import type { ExtractedInvoice, LineItem, PaymentMode, PdfSourceType } from "./models";

const MONTH_MAP: Record<string, string> = {
  jan: "01", feb: "02", mar: "03", apr: "04", may: "05", jun: "06",
  jul: "07", aug: "08", sep: "09", oct: "10", nov: "11", dec: "12",
};

const DATE_PATTERNS: Array<{ rx: RegExp; fmt: "dmy" | "ymd" | "dmy_mon" }> = [
  { rx: /(\d{2})[\/\-.](\d{2})[\/\-.](\d{4})/,                    fmt: "dmy" },
  { rx: /(\d{4})[\/\-](\d{2})[\/\-](\d{2})/,                     fmt: "ymd" },
  { rx: /(\d{1,2})\s+(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)[a-z]*\s+(\d{4})/i, fmt: "dmy_mon" },
];

const GSTIN_RX = /\b(\d{2}[A-Z]{5}\d{4}[A-Z]{1}[1-9A-Z]{1}Z[0-9A-Z]{1})\b/;
const INVOICE_NO_RX = /invoice\s*(?:no|number|#)\s*[:\-]?\s*([A-Z0-9][A-Z0-9\-\/]{3,29})/i;
// Order matters: more specific patterns first
const TOTAL_RX = [
  // "Invoice Value" label (Amazon, Flipkart)
  /invoice\s*value[:\s]*(?:₹|Rs\.?|INR)?\s*([\d,]+\.?\d{0,2})/i,
  // Standard grand total keywords
  /(?:grand\s*total|total\s*amount|net\s*payable|amount\s*due|amount\s*payable|payable\s*amount)\s*[:\-]?\s*(?:₹|Rs\.?|INR)?\s*([\d,]+\.?\d{0,2})/i,
  // "TOTAL: ₹tax ₹grandtotal" — two ₹ on same TOTAL line; take the last one
  /^total[:\s]+(?:₹|Rs\.?)?\s*[\d,]+\.?\d*\s+(?:₹|Rs\.?)\s*([\d,]+\.?\d{0,2})/im,
  // Last ₹ amount on a line ending with it
  /(?:₹|Rs\.?)\s*([\d,]+\.?\d{0,2})\s*$/im,
];
const LINE_ITEM_RX = /^(.+?)\s+(\d+(?:\.\d+)?)\s+(?:₹|Rs\.?)?\s*([\d,]+\.?\d{0,2})/gim;

// Lines to skip when looking for merchant name or line item names
const HEADER_SKIP = [
  "tax invoice", "bill of supply", "cash memo", "original for recipient",
  "duplicate", "invoice", "receipt", "page ", "gstin", "pan no", "cin no",
  "authorized signatory", "e-invoice", "irn", "ack no",
];

// Additional phrases that disqualify a line-item name match
const ITEM_HEADER_SKIP = [
  ...HEADER_SKIP,
  "description", "item name", "sl. no", "hsn code", "unit price",
  "net amount", "tax rate", "tax type", "tax amount", "total amount",
  "original for", "seller services", "retail india", "fulfillment",
  "customers desirous", "business account", "demand for payment",
];

export function extractDate(text: string): string | null {
  for (const { rx, fmt } of DATE_PATTERNS) {
    const m = rx.exec(text);
    if (!m) continue;
    if (fmt === "dmy") return `${m[3]}-${m[2]}-${m[1]}`;
    if (fmt === "ymd") return `${m[1]}-${m[2]}-${m[3]}`;
    if (fmt === "dmy_mon") {
      const mon = MONTH_MAP[m[2].toLowerCase().slice(0, 3)] ?? "01";
      return `${m[3]}-${mon}-${m[1].padStart(2, "0")}`;
    }
  }
  return null;
}

export function extractGstin(text: string): string | null {
  return GSTIN_RX.exec(text)?.[1] ?? null;
}

export function extractInvoiceNumber(text: string): string | null {
  return INVOICE_NO_RX.exec(text)?.[1]?.trim() ?? null;
}

export function extractMerchantName(text: string): string | null {
  // 1. "Sold By :" / "Seller:" label (Amazon, Flipkart, etc.) — name is on the next line
  const soldByM = /(?:sold\s+by|seller)\s*[:\-]\s*\n?\s*([A-Z][A-Za-z0-9\s,.()\-&']{4,80})/im.exec(text);
  if (soldByM) return soldByM[1].trim().replace(/\s+/g, " ");

  // 2. First line that contains a known company-type suffix
  const companyM = /^([A-Z][A-Za-z0-9\s,.()\-&']{3,70}(?:private\s+limited|pvt\.?\s*ltd\.?|(?<!\b\w)\blimited\b|llp|llc))/im.exec(text);
  if (companyM) return companyM[1].trim().replace(/\s+/g, " ");

  // 3. Fallback: first non-empty, digit-sparse line that isn't a known invoice header
  const lines = text.split("\n");
  for (const line of lines) {
    const t = line.trim();
    if (!t) continue;
    const tl = t.toLowerCase();
    if (HEADER_SKIP.some((w) => tl.includes(w))) continue;
    const digits = (t.match(/\d/g) ?? []).length;
    if (digits / t.length < 0.4) return t;
  }
  return null;
}

export function extractGrandTotal(text: string): number | null {
  for (const rx of TOTAL_RX) {
    const m = rx.exec(text);
    if (m) {
      const val = parseFloat(m[1].replace(/,/g, ""));
      if (!isNaN(val)) return Math.round(val * 100);
    }
  }
  return null;
}

export function extractPaymentMode(text: string): PaymentMode {
  const l = text.toLowerCase();
  if (l.includes("cash")) return "cash";
  if (l.includes("upi") || l.includes("gpay") || l.includes("phonepe")) return "upi";
  if (l.includes("card") || l.includes("credit") || l.includes("debit")) return "card";
  if (l.includes("bnpl") || l.includes("buy now") || l.includes("emi")) return "bnpl";
  return "unknown";
}

export function extractLineItems(text: string): LineItem[] {
  const items: LineItem[] = [];
  let m: RegExpExecArray | null;
  LINE_ITEM_RX.lastIndex = 0;
  while ((m = LINE_ITEM_RX.exec(text)) !== null) {
    const name = m[1].trim();
    const qty = parseFloat(m[2]);
    const unitPricePaise = Math.round(parseFloat(m[3].replace(/,/g, "")) * 100);
    if (!name || unitPricePaise <= 0) continue;
    // Reject obvious header rows and boilerplate blobs
    if (name.length > 200) continue;
    const nameLower = name.toLowerCase();
    if (ITEM_HEADER_SKIP.some((w) => nameLower.includes(w))) continue;
    items.push({
      name,
      quantity: qty,
      unitPricePaise,
      totalPricePaise: Math.round(qty * unitPricePaise),
      discountPaise: 0,
    });
  }
  return items;
}

export function computeConfidence(invoice: Partial<ExtractedInvoice>): number {
  let hits = 0;
  if (invoice.merchantName) hits++;
  if (invoice.invoiceDate) hits++;
  if (invoice.grandTotalPaise != null) hits++;
  if (invoice.paymentMode && invoice.paymentMode !== "unknown") hits++;
  return hits / 4;
}

export function parse(text: string, sourceType: PdfSourceType): ExtractedInvoice {
  const merchantName = extractMerchantName(text);
  const invoiceDate = extractDate(text);
  const merchantGstin = extractGstin(text);
  const grandTotalPaise = extractGrandTotal(text);
  const paymentMode = extractPaymentMode(text);
  const lineItems = extractLineItems(text);
  const invoiceNumber = extractInvoiceNumber(text);

  const partial = { merchantName, invoiceDate, grandTotalPaise, paymentMode };
  const confidenceScore = computeConfidence(partial);

  return {
    merchantName,
    merchantAddress: null,
    merchantGstin,
    merchantPhone: null,
    merchantPincode: null,
    invoiceNumber,
    invoiceDate,
    lineItems,
    subtotalPaise: null,
    discountPaise: 0,
    taxPaise: null,
    grandTotalPaise,
    paymentMode,
    sourceType,
    rawText: text,
    confidenceScore,
  };
}

File: iOS/Sources/Core/Extraction/InvoiceFieldParser.swift
import Foundation

enum InvoiceFieldParser {

    private static let datePatterns: [(NSRegularExpression, String)] = {
        let specs: [(String, String)] = [
            (#"(\d{2})[\/\-](\d{2})[\/\-](\d{4})"#, "dmy"),
            (#"(\d{4})[\/\-](\d{2})[\/\-](\d{2})"#, "ymd"),
            (#"(\d{1,2})\s+(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)[a-z]*\s+(\d{4})"#, "dmy_mon"),
        ]
        return specs.compactMap { pattern, fmt -> (NSRegularExpression, String)? in
            guard let rx = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive) else { return nil }
            return (rx, fmt)
        }
    }()

    private static let gstinRegex = try! NSRegularExpression(
        pattern: #"\b(\d{2}[A-Z]{5}\d{4}[A-Z]{1}[1-9A-Z]{1}Z[0-9A-Z]{1})\b"#
    )

    private static let totalPatterns: [NSRegularExpression] = [
        try! NSRegularExpression(pattern: #"(?:grand\s*total|total\s*amount|net\s*payable|amount\s*due)\s*[:\-]?\s*(?:₹|Rs\.?|INR)?\s*([\d,]+\.?\d{0,2})"#, options: .caseInsensitive),
        try! NSRegularExpression(pattern: #"(?:₹|Rs\.?)\s*([\d,]+\.?\d{0,2})\s*$"#, options: [.caseInsensitive, .anchorsMatchLines]),
    ]

    private static let monthMap: [String: String] = [
        "jan": "01", "feb": "02", "mar": "03", "apr": "04", "may": "05", "jun": "06",
        "jul": "07", "aug": "08", "sep": "09", "oct": "10", "nov": "11", "dec": "12",
    ]

    static func extractDate(from text: String) -> String? {
        let range = NSRange(text.startIndex..., in: text)
        for (rx, fmt) in datePatterns {
            guard let match = rx.firstMatch(in: text, range: range) else { continue }
            let groups = (1...match.numberOfRanges - 1).map { i -> String in
                if let r = Range(match.range(at: i), in: text) { return String(text[r]) }
                return ""
            }
            switch fmt {
            case "dmy":
                let (d, m, y) = (groups[0], groups[1], groups[2])
                return "\(y)-\(m)-\(d)"
            case "ymd":
                let (y, m, d) = (groups[0], groups[1], groups[2])
                return "\(y)-\(m)-\(d)"
            case "dmy_mon":
                let d = String(format: "%02d", Int(groups[0]) ?? 1)
                let m = monthMap[groups[1].lowercased().prefix(3).description] ?? "01"
                let y = groups[2]
                return "\(y)-\(m)-\(d)"
            default: break
            }
        }
        return nil
    }

    static func extractGstin(from text: String) -> String? {
        let range = NSRange(text.startIndex..., in: text)
        guard let match = gstinRegex.firstMatch(in: text, range: range),
              let r = Range(match.range(at: 1), in: text) else { return nil }
        return String(text[r])
    }

    static func extractMerchantName(from text: String) -> String? {
        let lines = text.components(separatedBy: .newlines)
        return lines.first { line in
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            guard !trimmed.isEmpty else { return false }
            let digits = trimmed.filter { $0.isNumber }
            return Double(digits.count) / Double(trimmed.count) < 0.5
        }
    }

    static func extractGrandTotal(from text: String) -> Int64? {
        let range = NSRange(text.startIndex..., in: text)
        for rx in totalPatterns {
            guard let match = rx.firstMatch(in: text, range: range),
                  let r = Range(match.range(at: 1), in: text) else { continue }
            let raw = String(text[r]).replacingOccurrences(of: ",", with: "")
            if let amount = Double(raw) {
                return Int64((amount * 100).rounded())
            }
        }
        return nil
    }

    static func extractPaymentMode(from text: String) -> PaymentMode {
        let lower = text.lowercased()
        if lower.contains("cash") { return .cash }
        if lower.contains("upi") || lower.contains("gpay") || lower.contains("phonepe") { return .upi }
        if lower.contains("card") || lower.contains("credit") || lower.contains("debit") { return .card }
        if lower.contains("bnpl") || lower.contains("buy now") || lower.contains("emi") { return .bnpl }
        return .unknown
    }

    static func extractLineItems(from text: String) -> [LineItem] {
        let pattern = try! NSRegularExpression(
            pattern: #"^(.+?)\s+(\d+(?:\.\d+)?)\s+(?:₹|Rs\.?)?\s*([\d,]+\.?\d{0,2})"#,
            options: [.caseInsensitive, .anchorsMatchLines]
        )
        let range = NSRange(text.startIndex..., in: text)
        var items: [LineItem] = []
        pattern.enumerateMatches(in: text, range: range) { match, _, _ in
            guard let m = match,
                  let nameR = Range(m.range(at: 1), in: text),
                  let qtyR = Range(m.range(at: 2), in: text),
                  let priceR = Range(m.range(at: 3), in: text) else { return }
            let name = String(text[nameR]).trimmingCharacters(in: .whitespaces)
            let qty = Double(text[qtyR]) ?? 1.0
            let priceStr = String(text[priceR]).replacingOccurrences(of: ",", with: "")
            let unitPaise = Int64(((Double(priceStr) ?? 0) * 100).rounded())
            guard !name.isEmpty, unitPaise > 0 else { return }
            items.append(LineItem(name: name, quantity: qty, unitPricePaise: unitPaise))
        }
        return items
    }

    static func computeConfidence(invoice: ExtractedInvoice) -> Float {
        var hits = 0
        if invoice.merchantName != nil { hits += 1 }
        if invoice.invoiceDate != nil { hits += 1 }
        if invoice.grandTotalPaise != nil { hits += 1 }
        if invoice.paymentMode != nil && invoice.paymentMode != .unknown { hits += 1 }
        return Float(hits) / 4.0
    }

    static func parse(text: String, sourceType: PdfSourceType) -> ExtractedInvoice {
        let merchant = extractMerchantName(from: text)
        let date = extractDate(from: text)
        let gstin = extractGstin(from: text)
        let total = extractGrandTotal(from: text)
        let mode = extractPaymentMode(from: text)
        let items = extractLineItems(from: text)

        let partial = ExtractedInvoice(
            merchantName: merchant,
            merchantAddress: nil,
            merchantGstin: gstin,
            invoiceDate: date,
            lineItems: items,
            subtotalPaise: nil,
            discountPaise: 0,
            taxPaise: nil,
            grandTotalPaise: total,
            paymentMode: mode,
            sourceType: sourceType,
            rawText: text,
            confidenceScore: 0
        )
        let confidence = computeConfidence(invoice: partial)
        return ExtractedInvoice(
            merchantName: merchant,
            merchantAddress: nil,
            merchantGstin: gstin,
            invoiceDate: date,
            lineItems: items,
            subtotalPaise: nil,
            discountPaise: 0,
            taxPaise: nil,
            grandTotalPaise: total,
            paymentMode: mode,
            sourceType: sourceType,
            rawText: text,
            confidenceScore: confidence
        )
    }
}
